### DISCUSSION TITLE
A constraint-based model of CatVars

### DISCUSSION THREAD

**Discussion by DanielPuthawala** [ga4gh/cat-vrs#22] [General]
The general approach to schemata in VRS, and by extension, the aspects that Cat-VRS inherited through VRSatile, has been to model different types of variants as discrete top-level data classes.  Hence, having one data class for canonical variants, another for protein consequence variants, a third for CNVs, etc.  There are, I think two potential issues with this approach in the categorical variation space: lots of significant (for clinical analysis and research) CatVars that fall into the cracks, and the need to compute entailments.

(The below rhetorical example is based on the CatCNV spec as per the [1.0.0.connect.2024-04.1](https://github.com/ga4gh/cat-vrs/releases/tag/1.0.0.connect.2024-04.1) snapshot release)

CatVars in the cracks: If we consider just CNVs and sequence variants for a moment, we can, and have created two different classes for these.  Sequence variants are in reference to a canonical allele, and some alternate sequence, while CNVs may also refer to some reference sequence and require some copy number or copy change value.  Other than the fact that they both in reference to a stand genome build, there is little overlap in terms of the data types used in this two classes.

Now consider expression/function variants.  These have a lot in common with sequence variants, and often contain information on some sort of sequence variation, such as a deletion, gene loss, stop gain, etc.  But they also require a quantitative data field that has much in common with that from CNVs, but in this case typically something related to levels of gene product.  We could make a third, discrete, and independent data class of variants that contains features from both sequence variants and CNVs, but I don’t like this for a couple reasons.  First, it’s not DRY.  Second, if a CatVar entry is made, say, as a sequence variant, and then it turns out that there is a functional consequence of that sequence change, this would require re-configuring the CatVar as a function variant, not just adding the new functional information, which doesn’t seem efficient.  And third, it doesn’t solve the problem.  If we make that third class, there will still be variants falling into the cracks between them,

Entailments: Several of our use cases so far involve matching between categorical variants, and a subset of that involves determining if a given CatVar partially intersects or is a subset of another CatVar.  If we have discrete, disparate variant classes, then we can explicitly code that CatVars in one class are subsets of those in another class, but making any other (eg orthogonal) comparisons become nontrivial given the apples-to-oranges nature of the data representations.

An alternative is to make a constraint-based model of categorical variation.  In this model, instead of the top-level objects being classes of CatVars, the top level objects here are constraints on variation, with many of these constraints being optional. The top-level CatVar classes in the previous model here instead correspond to profiles which specify certain constraints that must be satisfying in order to belong in that class.  Under this model, all CatVars are stored in a similar formal, just with varying degrees of optional constraints having non-null values.

This could solve the gaps problem, because even if a variant doesn’t fit into any given profile, it can still be represented as thoroughly or precisely as any other variant, as long as the spec has been built out enough to include the relevant constraint fields.

This also can solve the entailment/match issue, because computing the relationship between two given caters becomes an exercise of simply comparing their constraints on an apples-to-apples basis, since they all underlying have the same possible constraints fields.  I’ll sketch out a visual aid of these two approaches and add it below.

So, in summary, the proposal is to shift from modelling variant classes into modelling CatVar constraint classes, and that under the new model, conventional CatVar classes would correspond to profiles: lists of constraints with particular values indicative of CatVars belonging to that class.

**Comment by DanielPuthawala** [comment]
![Cat-VRS schemata - CatVar Class Model (1)](https://github.com/ga4gh/cat-vrs/assets/140105309/914d3cfc-8579-48f8-af0a-a170f517e13b)

This sketch shows 2 CatVar classes under the current model.

**Comment by DanielPuthawala** [comment]
![Cat-VRS schemata - CatVar Constraint Model](https://github.com/ga4gh/cat-vrs/assets/140105309/d1d43dd3-27ca-4371-9a81-45269b2c4428)

Under the constraint model, we have a set of constraints that are common to all CatVars.  Now, CatCNV and Protein Sequence Variants are profiles which specify core constraints needed for a CatVar of that profile to be queried.

**Comment by DanielPuthawala** [comment]
![Cat-VRS schemata - Example Function Variant](https://github.com/ga4gh/cat-vrs/assets/140105309/ea78ba7f-8cb1-4810-9c7f-828df3c34f12)

A Function CatVar, showing constraints with or without non-null data values, and how it **_is_** a sequence variant by matching the required constraints of that profile, but we are also able to capture additional information that is more CNV-like, here being a relative change in sensitivity (here I'm deliberately construing "copyChange" as being generalizable to relative quantity changes.)

**Reply by brendanreardon** [reply]
Pairing this with Alex's comment [below](https://github.com/ga4gh/cat-vrs/discussions/22#discussioncomment-9547661), are your labels for Profile 1 and 2 ...
1. Placeholders for specific CatVars (e.g., _PIK3CA_ p.E545) or
2. Defining "sets" of CatVars that at least have non-null values for, at least, constraints connected by solid lines
?

**Comment by ahwagner** [comment]
So, as a general model, the constraints are applied following this general pattern?

![Cat-VRS Constraint Model (1)](https://github.com/ga4gh/cat-vrs/assets/811065/e9032bf7-830c-4f99-9503-5600dbd4fa9d)

**Reply by DanielPuthawala** [reply]
Yes, that's accurate!

**Comment by larrybabb** [comment]
This seems like a very reasonable approach to take. My biggest concern is whether this approach should be pursued in a 2.x release path because this change is a breaking change with vrsatile which our current 1.x model is based on. If we want this to be our 1.x direction I will need a very quick set of defined constraints for canonicalallele, categorical in, and variantdescriptor (text). Otherwise this will set the GKS pilot work back many months at a time where we have just finally gotten a 1.x pipeline set up after years of trying. That's a tough sell for me.

It comes down to how quickly we can draft a viable constraint schema and demonstrate how we discern one subtype of constraints from another. Will there be some standard and naming of constraint profiles? Can we put some sample data together from real world data really quickly. If not then let's pursue this as 2.x and stick with 1.x as is.

**Reply by DanielPuthawala** [reply]
Larry, Thanks for that feedback!  I don't want to set you back, and I think that this hopefully will amount to a fairly small update on your end.  (See below for some demos). As for the version numbering, I am less concerned about the 1.x vs 2.x distinction (because versioning is now both a function of breaking changes the maturty of the affected data classes) and more concerned with getting something to you that (a) won't be a pain to adopt if and when you're ready and able to, and (b) is compatible with the general model structure we anticipate needing so we hopefully don't have to overhaul it dramatically in the future.

**Comment by larrybabb** [comment]
I'd like to understand if we would create subclasses of the abstract constraint class for the various concrete concepts needed. Or if we would leave it to users to inspect a given constraint object to determine what type of catvar it actually is?

**Reply by DanielPuthawala** [reply]
That's the neat thing about the constraint model, you can do both/either.  You can define custom profiles for whatever major CatVars you are interested in, but at the same time, this structure should allow us to parse out CatVar types on the fly, so you can determine exactly what type a given CatVar is (and determine if another is the same type/in the same class or not) on the fly.

**Comment by brendanreardon** [comment]
Thank you for laying this out so thoughtfully! I think the 4 figures currently presented in this discussion will be really wonderful visual aids in the next call. This is a very nice framework. I wonder if some of Larry's concerns come from the [Connect release](1.0.0.connect.2024-04.1) being labeled as 1.0.0, instead of leaving the major version to be 0.

**Reply by DanielPuthawala** [reply]
The connect snapshot release is just that, a snapshot so we have a point-in-time reference of where we were at connect.  And AFAIK we need to start with at least a 1.x.x.  

I'm assuming that we will version in lie with the proposed GKS maturity model (@ahwagner is that fit for public review/circulation yet?) but the basic gist is that all of the data classes are currently draft maturity status.  updates to just draft-status classes entail a patch (1.0.1, 1.0.2, 1.0.3, etc).  Moving classes up to trial-use maturity or making breaking changes to trial-use-status classes requires a minor release (1.1.x, 1.2.x, etc), and this is the point in which we will need PRC review and steerign approval to official launch or 'release' Cat-VRS as a GA4GH GKS product.  

So Larry's concern seems more that he wants us to keep things unchanged for 1.1.0, and then update for 1.2.0, but I'm not sure that makes sense to submit what we started with for PRC review and approval, rather than what we're actually developing.  But at the same time I don't want to cause him a huge headache.  Hopefully the examples below show little would be required to be tweaked on Larry for CatVars of the existing `canonicalAllele` and `categoricalCnv` classes.

**Comment by DanielPuthawala** [comment]
@larrybabb Here's some examples of what updating a `canonicalAllele` CatVar and `categoricalCnv` CatVar would entail to make them compatible with a constraint model (borrowing the visual language of Alex's diagram).  I am eager to hear your thoughts!

![Cat-VRS Constraint Model - Can Allele demo 1](https://github.com/ga4gh/cat-vrs/assets/140105309/3dde1e41-6179-475d-bb11-ab4294500321)

![Cat-VRS Constraint Model - Cat CNV Demo 1](https://github.com/ga4gh/cat-vrs/assets/140105309/bc3ee6cd-806a-47fe-92b6-554480a06946)

**Reply by ahwagner** [reply]
For the last example; shouldn't the `members` be VRS `CopyNumberCount` objects, not `SequenceLocation`?

**Reply by DanielPuthawala** [reply]
That would make sense, but I just [pulled this from this pull from Clinvar](https://raw.githubusercontent.com/ga4gh/gk-pilot/main/datasets/clinvar/catvars.json), so that's on @larrybabb 's end?

**Comment by larrybabb** [comment]
Thanks @DanielPuthawala for putting together this side-by-side example.  

While I agree with the comments above that this is a very flexible and extensible framework for defining all types and kinds of constraints to form various sets of variants into different `Categorical` forms, I'm not sure that is what we the cat-vrs specification is intended to do.

I'd like to the various techniques that a receiving system would have to engineer in order to discern what type of categorical variant they are receiving.  If they are all explicitly typed as "CategoricalVariation" then I believe there may be less value in sharing it other than to generalize what is being received.

In VRS we wrestled with the idea of `Typing` the messages explicitly so that it would be possible for consuming systems to both validate the data type and have sufficient information to process it.

I suppose having a generalized schema is fine, but the specificity will come in terms of putting requirements on the constraints themselves. We are essentially pushing the issue of typing into the constraint nodes specification itself. I think this is a fine idea, but one that needs to be done a part of the definition of the various categorical variation that we want to define.

Optionality in a specification is great as long as it does not redefine a new concept.  Getting the set of concepts that are comparable and contrastable such that they can be reasonably implemented by a system is the key.  I understand that it is feasible to create a generalized model and then claim that by applying the constraints of two differently defined set of constraints you can still find the overlapping members, I'm not sure that is what we need to do.  

What we need is the ability to figure out what the members are in a pragmatic way so that we can discern one instance of a categorical variant from another for the same set of constraints (aka. the categorical variant subtype). Then we can assign identifiers for these subtypes where the subtype itself acts as a kind of namespace.  I believe this is a more sensible approach than thinking about all categorical variation as one thing.

**Reply by DanielPuthawala** [reply]
> What we need is the ability to figure out what the members are in a pragmatic way

While I agree with you that CatVars are fundamentally sets of assayed variation, the problem is that in many cases, the possible cardinality of those sets is astronomically large.  So determining if two CatVars are the same by lining up their members and seeing if they're the same simply isn't feasible.  Suppose we have the CatVar "Insertions of <135 nucleotides in [POSITION]" that set has 4^134 possible members, roughly equivalent to [the estimated number of atoms in the observable universe](https://www.livescience.com/how-many-atoms-in-universe.html).

That's why Alex and I are defining the core of these CatVars not by their members, but the properties common to all their members.  It is way faster and easier than checking if it matches one of the members of that set. 

If you DO want to compare against members for really small sets or for sets with a couple really common or representative members, well that's what leaving the optional field in there is for!

**Comment by larrybabb** [comment]
One other observation.  Maybe we need to simply define a "VariationConstraint" class that can be subclassed into the various forms, like CanonicalAllele (genomic defining context) or CanonicalLocation (genomic defining location) or ProteinExpression, ... etc..

Then a system can be engineered to reference the Constraint type to now when they are dealing with one type of categorical variation versus another?  

Just a thought.  I'm not sure this makes things more complex vs just putting subtypes on the CategoricalVariation itself (if we presume all different constraint types will share the same exact set of optional attributes (members, expressions, etc..).

**Comment by brendanreardon** [comment]
I have some time over the next week to work on some examples for the constraint model, as we discussed during last week's call. Can one of you point me to the current specifications for canonicalAllele and categoricalCNV? For sequence variants, the [VRS documentation](https://vrs.ga4gh.org/en/stable/terms_and_model.html?highlight=copy#) is robust plus there are examples in both the [va-spec](https://github.com/ga4gh/va-spec/tree/1.x/examples) and [gk-pilot](https://github.com/ga4gh/gk-pilot/tree/main/datasets) repositories. 

@DanielPuthawala if this would be easier as a call, we can definitely do that too 🙈 !

**Reply by DanielPuthawala** [reply]
I'm down to work with you, Brendan!  We can coordinate a time offline here.

The most up-to-date [ReadTheDocs is here](https://cat-vrs.readthedocs.io/en/latest/), though this is running behind t[he current dev branch](https://github.com/ga4gh/cat-vrs/tree/Issue-14).  If you want to see what is currently being worked on (not that I've made much progress in the npast week), you can find that [in the code base here](https://github.com/ga4gh/cat-vrs/blob/Issue-14/schema/catvrs/catvrs-source.yaml).

**Comment by DanielPuthawala** [comment]
Here's two more mock-ups that are a bit more exploratory.

This first demo is for a protein sequence variant, specifically for a variation in amino acid sequence, rather than being predicted from a nucleotide sequence variant.  This example is adapted from variant `627218 ` from the same Clinvar dataset linked above.  As a sequence variant, it has much in common with the canonical allele example earlier.

![Cat-VRS Constraint Model - Prot Seq Cons Demo 1](https://github.com/ga4gh/cat-vrs/assets/140105309/e08cc6d1-e597-4323-9ff2-01efaaf5498b)

This second mockup is for a functional variant, here for ["BRCA1 Loss-of-function", adapted from this record in CIViC](https://civicdb.org/variants/131/summary).  This mockup incudes an additional `functionConstraint` which functions similarly to the `copyConstraint` in [the categoricalCNV example above](https://github.com/ga4gh/cat-vrs/discussions/22#discussioncomment-9649033).

![Cat-VRS Constraint Model - Functin Var Demo](https://github.com/ga4gh/cat-vrs/assets/140105309/3da1fb51-0524-476f-b497-71d6a9be3ed8)

**Reply by DanielPuthawala** [reply]
I should add that the italicized bits, like _"Allele"_ for the defining context and _"123456789_ for the code are simply placeholders!

**Comment by DanielPuthawala** [comment]
I've made a minor tweak to the Protein Sequence Consequence variant demo and some more major revisions to the Function Variant demo.

The PSC demo was just tweaked to include relations that make more sense for a variant label that denotes a amino acid sequence change.  So rather than something like `genomeLiftover` we have an example mapping like `codonAlignment`.

![Cat-VRS Constraint Model - Prot Seq Cons Demo 1 (1)](https://github.com/ga4gh/cat-vrs/assets/140105309/817989bf-4461-4cad-a3e9-da1aa34d45bf)

The Function Variant demo has been tweaked more extensively.  

First, given that `vrs:allele` described nucleotide location-states, and functional variants pertain more to protein function, it seems reasonable to assume that the context of a function variant is a gene, rather than a `vrs:allele`.

Second, the `functionConstraint` has been updated.  First, the values of `functionalConsequence` has been updated to reflect [Muller's morphs typology of gene product function changes](https://en.wikipedia.org/wiki/Muller%27s_morphs).  These values have also been updated with the closest SO codes.  For most, there is a close match, such as `hypermorphic` corresponding to ["SO:0001557 polypeptide_gain_of_function_variant"](http://www.sequenceontology.org/browser/current_release/term/SO:0001557).  For `amorphic` and `neomorphic`, there is not an exact protein-specific concept in SO (that I can find), so I've gone with the closest match: "
[SO:0002054 loss_of_function_variant](http://www.sequenceontology.org/browser/current_release/term/SO:0002054) for `amorphic`, and ["SO:0002053 gain_of_function_variant"](http://www.sequenceontology.org/browser/current_release/term/SO:0002053) for `neomorphic`.

Does this seem reasonable?  Is there a better typology of protein function?   Is there a better-fitting ontology we should reference than SO? (OR should I talk to SO about getting codes added for `amorphic` and `neomorphic`?)

Third, since I have removed the `quantityConstraint` from being a property of the `functionConstraint` and made it into a constraint in it's own right, which may very well being the same as the `copyConstraint` from the [categoricalCNV demo previously](https://github.com/ga4gh/cat-vrs/discussions/22#discussioncomment-9649033).  Haven't gotten there yet.

![Cat-VRS Constraint Model - Function Var Demo](https://github.com/ga4gh/cat-vrs/assets/140105309/4a413137-ca94-4604-b010-6d88fa54bd37)

Hope to hear your thoughts.  I'd like to discuss this at our meeting on 7/3/24

**Reply by brendanreardon** [reply]
This is great and will be nice aids for a discussion on Wednesday. Thank you!! Some comments that I have are...

Naive question - I've been unable to find codonAlignment, genomeLiftover, transcriptProject, or any permutations of them in sequence ontology. Are these proposed terms for additions to a future release? 

> First, given that vrs:allele described nucleotide location-states, and functional variants pertain more to protein function, it seems reasonable to assume that the context of a function variant is a gene, rather than a vrs:allele.
I think that this is reasonable... Is the correct interpretation that the "BRCA1 loss-of-function" would be a standalone catvar, and then it would have members of specific _BRCA1_ variants that cause loss-of-function; e.g., [NM_007294.4(BRCA1):p.Tyr1853Ter](https://gnomad.broadinstitute.org/variant/17-43045711-G-GT?dataset=gnomad_r4)? And then the catvar for NM_007294.4(BRCA1):p.Tyr1853Ter would include the functionConstraint as an optional constraint? 

I agree that those are appropriate matches in sequence ontology for Muller's morphs! And I agree that quantity and function should be separate constraints, especially since we are defining the minimum set for each categorical variant type. I'm looking forward to the discussion on Wednesday! Thank you, Daniel!

**Reply by DanielPuthawala** [reply]
Ah!  The relations aren't from SO, I was just looking for equivalent coded terms for the functional consequence constraint.  I've been mocking up the relations as methods to be used at the whim of the implementer.

**Reply by egchristensen** [reply]
I believe these terms fall in scope of SO as long as we are trying to represent these concepts at the genome level or at the sequence level using genomic coordinates. As far as I know, there isn't another ontology that aims to represent the sequence features in the way that SO does. There may be some orphan ontologies or other spinoffs, but SO is the main value set for sequence annotations as far as I know. RefSeq and GENCODE/Ensembl annotation are really the two main sources for human genome annotations in open science and they both use SO terms to transmit their sequence annotations to the public. The most recent effort to sort out variation terminology at SO was in 2015 when a previous PhD student (Nicole Ruiz-Schultz) went through and did some work. See figure 1 of ["Improving the Sequence Ontology terminology for genomic variant annotation"](https://pubmed.ncbi.nlm.nih.gov/26229585/). Many domains of variation terminology in SO still need work, so we're open to suggestions. Generally, SO prefers to model representations at the genome level since that's the most central point of reference we have for sequence annotations. This has implications for the representation of variation (especially structural variation) since not all types of variation lend themselves to being represented on a linear genome with genomic coordinates.  

If you had to group SO terms into three (imperfect) groups, you could generally categorize them as either **a feature, a change, or an effect.** A feature is a region of sequence (or the space between two bases) like an exon, a CDS, or a gene region. A change encompasses most types of variation or sequence comparisons. An effect is the result of that change, typically in terms of the morphology or function of a gene product. Admittedly, SO is not always consistent with the feature/change/effect categories, but I'm hoping to get that sorted before I graduate. 

My question when it comes to Muller's typology is not whether the terms are descriptive enough on their own, but rather is Muller assigning types/categories to sequence features, changes, or effects? The answer here affect what part of SO we want to edit. When we represent Muller's types on the genome, should we simply assign a region of the genome to one of Muller's types and somehow associate that with the coding region (CDS) or should we maybe make Muller's type an attribute of the CDS annotation itself? It seems that Muller's categories would fall under the effect category. For example, some type of variation (insertion, deletion, frameshift, etc.) could result in a hypermorphic, antimorphic, or a neomorphic gene product? The phrase _"A deletion from coordinate_1 to coordinate_2 resulted in hypomorphic gene product causing a_partial_loss_of_gene_function"_ makes more sense to me than the phrase _"A hypomorphic deletion from coordinate_1 to coordinate_2 caused a_partial_loss_of_gene_function"_. The former treats Muller's type as an effect category rather than a change category. 

Constraining sequence features to certain biological types of sequence (DNA, RNA, AA, etc.) is a major factor to consider when representing transcriptome-specific and proteome-specific annotations at the genome level. _This is also relevant to the contextualization of any types of variation that appear within sequence regions that have been constrained to certain types of biological sequence._ Ensembl's solution is to associate biotypes with sequence annotations, but Ensembl's biotype entity has become conceptually overloaded. Biotypes at Ensembl no longer just represent biological types of sequence, but they've also come to represent various stages of the flow of genomic information, stages of knowledge curation, and experimental evidence (the sort of stuff that VA is working on). [Here's a list of Ensembl Biotypes from human genome build 37](https://grch37.ensembl.org/info/genome/genebuild/biotypes.html). RefSeq also has a similar biotype object, but they don't use it to the same extent as GENCODE/Ensembl because the data models in place at NCBI and EMBL-EBI essentially take opposite approaches to modeling this information.

I think SO will work for this use case, but we'll need to dig a little deeper to make sure the ontology has the right terms and relations to enable the sort of the logic that you're all trying to represent (at the genome level not the sequence level). I have a deadline of August 15th to finish and write up my review of sequence annotation data models. Let me wrap that up and then maybe we could strategize a way forward with SO if you're interested?

**Reply by DanielPuthawala** [reply]
Small update to Function constraint with `functionTarget` property.
![Cat-VRS Constraint Model - Function Var Demo (1)](https://github.com/user-attachments/assets/8df013fc-7516-4758-828a-620fbc9e726f)

**Reply by brendanreardon** [reply]
Per the discussion today reflecting on what function may or may not be out of scope for CatVRS, and instead may be in Variant Annotation. I _think_ that these items from the [Test Set Variants](https://docs.google.com/document/d/1aV-SqxdmuRN_EKvafzTSe0GoGC9yOzPsjrdWE0LXqYc/edit?usp=sharing) would be in scope: 
- Gain of function
- Loss of Function
- Wild type

And the following would be out of scope
- Conserved Domain Mutation
- Promoter Mutation

Uncertain
- Loss of Heterozygosity (LoH)
- Exon Skipping (e.g., MET Exon 14 skipping)

I agree that having oncogenicity and pathogenicity as being out of scope and moreso being for Variant Annotation makes sense. By this logic, LoH and Exon skipping would also be out of scope, I think, but I could be persuaded either way.

**Reply by DanielPuthawala** [reply]
Hey @brendanreardon Can you elaborate on the Out-of-scope and uncertain ones?

Promoter Mutation simply seems like a fairly benign variant to me, a location, in this case, some gene's promoter region, and a mutation, or minimally specific sequence variant.  Is this label specifically an annotation in some form?  Conserved Domain Mutation I can see a bit more, in that the location (which domains are evolutionarily conserved?) is itself the result of a lot of hard looking and analysis at the gene phylogenetically.  But that being said it's just some sort of variant at that region, no?

I can see a argument for LoH being excluded, but then it's my impression that that too is used as a categorical variant in these cases.  So for example, ["PTCH1 loss of heterozygosity was associated with increased duration of therapy"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4534527/)

**Reply by DanielPuthawala** [reply]
> Small update to Function constraint with `functionTarget` property. ![Cat-VRS Constraint Model - Function Var Demo (1)](https://private-user-images.githubusercontent.com/140105309/361510607-8df013fc-7516-4758-828a-620fbc9e726f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5NjExMDIsIm5iZiI6MTcyNDk2MDgwMiwicGF0aCI6Ii8xNDAxMDUzMDkvMzYxNTEwNjA3LThkZjAxM2ZjLTc1MTYtNDc1OC04MjhhLTYyMGZiYzllNzI2Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyOVQxOTQ2NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNmQ2N2Y1NjhlZTJmZWNmMjE2MDNkYWViYjU2NjRhZDEyNjhhYTZjMDcwZWNlNjZjZDc5ZDMyOTIzMmZhYjI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fjScf4WHQA21fasD82mDUveAnWb9ayLgTNn0_W_8Iuo)

In the 8/27 meeting, it was decided to remove this field on grounds that this probably spills over the line into VA territory, and is therefore probably not in-scope for Cat-VRS.

**Reply by brendanreardon** [reply]
When thinking about Promoter Mutations, I was considering the cases of annotating with a database such as [ENCODE](https://www.encodeproject.org/) to interpret if a variant is within a promoter region. Leaving it in the category of location though is totally fair, similar to describing a variant by the exon that it is located in.

What you wrote about the Conserved Domain was my exact rationale for it and Promoter Mutations. Would a delineation be...
- Alignment with standards and/or other data to categorize a variant - that would be out-of-scope of CatVRS and would be more for the variant annotation spec. For example, labeling NM_198253.3(TERT):c.-57A>C as a promoter mutation.
- Attaching a constraint when describing a categorical variant, and thus requiring that ... other variants meet that constraint to match. For example, making a categorical variant for TERT promoter mutations with such a constraint. 

... maybe I am over complicating the promoter use case because it could be inferred from a variant being in an untranslated region and the cDNA numbering 😵‍💫 

As for LoH and exon skipping, the examples above of categorizing individual variants versus applying a constraint (or, in other words, requiring that the condition of LoH or exon skipping is met) captures why I'm uncertain about them, I think. What do you think?

**Reply by DanielPuthawala** [reply]
I think we're on roughly the same page the way you captured it there.  

Deciding if something is a LoH or promoter mutation, and recording the information associated with that classification?  Yea, that's VA stuff.

Representing that there is a class of variants called "promoter mutations" or "LoH" variants in an efficiently computable fashion?  That's Cat-VRS.

So I reckon that even though there's definite overlap in the domains of variation that Cat-VRS and VA is going to count as in-scope, the purpose for those standards and the kind of information being recorded are going to be very different.

**Comment by brendanreardon** [comment]
As an FYI to myself (since I am still not entirely comfortable with these terms) and folks looking at these models, Daniel (I assume?) has drafted the [information model within the readthedocs](https://cat-vrs.readthedocs.io/en/latest/schema.html#categorical-variation-schemata) site for this project. Larry added that this schema, as used in the [Genomic Knowledge pilot](https://github.com/ga4gh/gk-pilot/tree/main) builds off of the [VRSATILE information model](https://vrsatile.readthedocs.io/en/latest/value_object_descriptor/index.html) ([VRSATILE catvars json](https://github.com/ga4gh/vrsatile/blob/gk-pilot/schema/catvars.json), thank you, Larry!!). 

<img width="661" alt="Screenshot 2024-07-16 at 10 11 38 AM" src="https://github.com/user-attachments/assets/166028e8-7b56-4112-93e4-dab5a0acf6cb">

_Current description of the information model in the [CatVar read the docs](https://cat-vrs.readthedocs.io/en/latest/schema.html#categorical-variation-schemata)_

Do people feel that these are fair descriptions of each field?
- **alternativeLabels** (list): List of strings of alternative / alias labels. 
    - This field _may_ be subject to change within a given system. 
- **constraints** (list): List of records that contain conditions which must be satisfied for a variant or alteration to match the described categorical variant. 
    - The constraints should **not** be updated for a given **id**, instead the record should be deprecated and a new one created.
- **description** (string): A free-text description of the entity. 
    - This field _may_ be subject to change within a given system.
- **id** (string): The logical, unique identifier of the entity in the system record in the string format of `{system}:{id}` such as `clinvar:999`. 
    - This id should **not** be updated for a given **constraint**, instead the record should be deprecated and a new one created.
- **extensions** (list): System-specific representation of a given categorical variant. 
    - This field _may_ be subject to change within a given system (?)
- **label** (string): Free-text label for the categorical variant, as used by the hosting system. Implementers are encouraged to follow standards (e.g., HGVS) for this field.
    - This field _may_ be subject to change within a given system. 
- **mappings** (list): A list of mappings to concepts in terminologies or concepts in **other** systems. Each record should contain the fields for "system" (the referenced system), "relation" (exactMatch, relatedMatch, or closeMatch - link to this specification?). 
    - This field _may_ be subject to change within a given system.
- **members** (list): A non-exhaustive list of ~~categorical~~ contextual variants that satisfy the constraints defined for this categorical variant's **constraints** (as represented within the same system?). 
    - This field _may_ be subject to change within a given system.
- **type** (string): "Categorical Variant"

One question that I have is:
- When should another categorical variant be placed in members instead of mappings, and mappings instead of members? I _think_ that the key differentiators are that members (1) satisfy the same constraints and (2) are within the same system, while mappings can be anything in other systems. 

p.s. I can move this into a separate discussion, but I thought that it would be appropriate here to help interpretation with the screenshots of the constraint-based model.

**Reply by larrybabb** [reply]
members are `contextual variants` not `categorical variants`,  so other repos that had `related` identifiers to their `categorical variants` would go in the `mappings` array.  The `members` would be derived, precise, `contextual variants` that are represented by the VRS schemas.

**Reply by DanielPuthawala** [reply]
Larry is your `contextual variant` here what I usually refer to as an `assayed variant`? ie a discrete measurable variant as seen in some given patient?

**Reply by larrybabb** [reply]
@brendanreardon If the table above is meant to depict the VRSATILE schema that I showed then it is missing the very important `definingContext` field that contains the key constraint for a `CanonicalAllele`.  If it is meant to be the newer Cat-Vrs schema then I think there should be a `constraints` field that is an array of some kind that could potentially contain the `definingContext` that is the source for all derived members.  But I would verify that second assumption with @DanielPuthawala

**Reply by larrybabb** [reply]
@DanielPuthawala the `contextual variant` can be referred to as the `assayed variant` or the `called assayed variant`, since it is reasonable to assume all `assayed variant calls` would be `contextual variants`. But a `contextual variant` is really just a variant that can be expressed using VRS.  It is a variant that is defined on a unique sequence at a unique location with a unique state (if it's an allele) or unique copy count (if its a copynumbercount) or a unique copy change (if it is a copynumber change), etc...
I've also heard folks refer to `assayed` or `contextual` variants as `observed` variants as well.

**Reply by DanielPuthawala** [reply]
I believe that that's correct, @larrybabb. 

Also the RTD schema at the moment is considerably behind the development and discussion of the constraint model.  (and I think is tied to the main 1.x branch.  I need to update it.

**Reply by DanielPuthawala** [reply]
Awesome, great.  Thank really clarifies a lot @larrybabb.  I belive I'm fully in agreement with your above reply to @brendanreardon.

**Reply by brendanreardon** [reply]
Thank you both!!

**Comment by DanielPuthawala** [comment]
Hey all, here's a long overdue update to the categoricalCNV demo, updated after conversation with Beth Pitel.
This version fleshes out the copyConstraint, to include information pertaining to CNVs as CatVars that are often referenced in the literature, FDA indications, etc, and I'd love to get your collective thoughts on this.  

Change notes:

- Fleshed out the copyConstraint.
- added the "scale" property in addition to the quantity property on the grounds that CNVs are often reported in terms of different reference scales, including the absolute number of copies, as well as relative copy changes in terms of being an x-fold change, or as the log2 value of the relative change.
- "copychange" started out as something a bit different, but ended up looking almost exactly like what's already in VRS, so it's quite possible we could just ref: that and simplify things a bit.
- added the amplificationThreshold as this is a property whose conventional value changes depending on the gene in question, or may specified on an ad-hoc basis.

![Cat-VRS Constraint Model - Cat CNV Demo 1 (1)](https://github.com/user-attachments/assets/8b7f8b10-1287-48b4-a082-c0a640ac49f2)

**Reply by brendanreardon** [reply]
Amazing! My only comment, I think, is that it may be worthwhile to generalize "amplificationThreshold" to be "threshold" to allow for thresholds being set for copy deletions. When fold change or log2values are reported for copy number segments, there usually is a specified threshold that a group uses to call low-level loss (sometimes also called shallow deletions) from loss or high-level loss. 

This is often done with segmented copy number from next-generation sequencing on total, or non-allele specific, copy number. For example, as discussed in [this post on the DepMap (cell line data) forums](https://forum.depmap.org/t/defining-deep-deletions-and-amplifications/710/4) and in the [GENIE data guide](https://www.aacr.org/wp-content/uploads/2022/03/GENIE_data_guide_11.0-public.pdf) for assessing copy number. The relevant bit for log2 ratio thresholds used by GENIE is on page 12 and reads:
> Thresholds for identifying gains and losses were set to 0.4 and (-)0.5, respectively; these values were obtained by examining the distribution of log2 ratios to identify peaks associated with copy number states. 

p.s. Beth ROCKS! 🪨

**Reply by DanielPuthawala** [reply]
Oh, and also, the Value in quanity should be 1, and not -1, since the scale is "absoluteCopyNumber".   -1 copies of a gene would be... a very interesting clinical case!

**Reply by DanielPuthawala** [reply]
Updated version of the CatCNV demo with these fixes and tweaks.
![Cat-VRS Constraint Model - Cat CNV Demo 1 (2)](https://github.com/user-attachments/assets/0ecf7c49-b7b3-4ad1-9edc-ab88b12d40bd)

**Reply by DanielPuthawala** [reply]
New updated version of the above, with `samplePurity` pulled out, and if needed, to be it's own constraint.
![Cat-VRS Constraint Model - Cat CNV Demo 1 (3)](https://github.com/user-attachments/assets/93d929b4-2efe-4a2f-88c3-69d928bcf8e0)

**Reply by Mrinal-Thomas-Epic** [reply]
This is very cool! I'm looking forward to seeing what the ClinVar data looks like when using this. I also have a couple questions (that maybe could be discussed at the next CatVRS call). Sorry if they have already been discussed earlier and I missed it.

1. How do you envision downstream software using the `log2Value`? Currently, there is no place in VRS to represent a `log2Value` (do we need to add this to VRS or another GKS area?). I'm not sure if we would be able to compare a VRS `CopyChange` or `CopyCount` to a `log2Value` threshhold.
2. How is a `threshold` different from a `quantity`? Does the `threshhold` need to be a string (rather than integer or range)?
3. Users of CatVRS shouldn't be able to provide combinations of constraints that aren't logically valid together. Does moving away from having separate classes for different types categorical variants (i.e., just `CategoricalVariant` instead of `CategoricalCNV`) break this principle, or do you have another way to handle this?

**Comment by ahwagner** [comment]
The constraint model has been implemented in #50.
